core/optaplanner-constraint-streams/src/main/java/org/optaplanner/constraint/streams/bavet/bi/JoinBiNode.java

```diff
@@ -0,0 +1,224 @@
+import java.util.LinkedHashSet;
+import java.util.Map;
+import java.util.Queue;
+import java.util.Set;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+import org.optaplanner.constraint.streams.bavet.common.AbstractNode;
+import org.optaplanner.constraint.streams.bavet.common.BavetTupleState;
+import org.optaplanner.constraint.streams.bavet.common.index.Indexer;
+import org.optaplanner.constraint.streams.bavet.tri.JoinTriNode;
+import org.optaplanner.constraint.streams.bavet.uni.UniScorer;
+import org.optaplanner.constraint.streams.bavet.uni.UniTuple;
+
+public final class JoinBiNode<A, B> extends AbstractNode {
+
+    private final Function<A, Object[]> mappingA;
+    private final Function<B, Object[]> mappingB;
+    /**
+     * Calls for example {@link BiScorer#insert(BiTuple)}, {@link JoinTriNode#insertAB(BiTuple)} and/or ...
+     */
+    public final Consumer<BiTuple<A, B>> nextNodesInsert;
+    /**
+     * Calls for example {@link BiScorer#retract(BiTuple)}, {@link JoinTriNode#insertAB(BiTuple)} and/or ...
+     */
+    public final Consumer<BiTuple<A, B>> nextNodesRetract;
+
+    private final Indexer<UniTuple<A>, Set<BiTuple<A, B>>> indexerA;
+    private final Map<UniTuple<A>, Object[]> indexPropertiesMapA = new HashMap<>();
+    private final Indexer<UniTuple<B>, Set<BiTuple<A, B>>> indexerB;
+    private final Map<UniTuple<B>, Object[]> indexPropertiesMapB = new HashMap<>();
+    private final Queue<BiTuple<A, B>> dirtyTupleQueue;
+
+    public JoinBiNode(Function<A, Object[]> mappingA, Function<B, Object[]> mappingB,
+            Consumer<BiTuple<A, B>> nextNodesInsert, Consumer<BiTuple<A, B>> nextNodesRetract,
+            Indexer<UniTuple<A>, Set<BiTuple<A, B>>> indexerA, Indexer<UniTuple<B>, Set<BiTuple<A, B>>> indexerB) {
+        this.mappingA = mappingA;
+        this.mappingB = mappingB;
+        this.nextNodesInsert = nextNodesInsert;
+        this.nextNodesRetract = nextNodesRetract;
+        this.indexerA = indexerA;
+        this.indexerB = indexerB;
+        dirtyTupleQueue = new ArrayDeque<>(1000);
+    }
+
+    public void insertA(UniTuple<A> tupleA) {
+        Object[] indexProperties = mappingA.apply(tupleA.factA);
+        Object[] old = indexPropertiesMapA.put(tupleA, indexProperties);
+        if (old != null) {
+            throw new IllegalStateException("Impossible state: the tuple for the fact ("
+                    + tupleA.factA + ") was already added in the indexPropertiesMapA.");
+        }
+
+        Map<UniTuple<B>, Set<BiTuple<A, B>>> tupleABSetMapB = indexerB.get(indexProperties);
+        // Use standard initial capacity (16) to grow into, unless we already know more is probably needed
+        Set<BiTuple<A, B>> tupleABSetA = new LinkedHashSet<>(Math.max(16, tupleABSetMapB.size()));
+        indexerA.put(indexProperties, tupleA, tupleABSetA);
+
+        tupleABSetMapB.forEach((tupleB, tupleABSetB) -> {
+            // TODO What if tupleB is dying?
```

Comment:
Indeed. Perhaps this PR has more work to do?

Reply:
This todo worries me. I 'll create a test to draw it out. Neither NQueens nor VRP would trigger if the bug exists, which means this PR might introduce a regression.